src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -1050,5 +1051,64 @@ private static TextEdit BuildRevertedEdit(TextEdit provisionalTextEdit)
             var formattingOptions = _formattingOptionsProvider.GetOptions(document.Uri);
             return Task.FromResult(formattingOptions);
         }
+
+        public override async Task<WorkspaceEdit?> RenameAsync(DelegatedRenameParams request, CancellationToken cancellationToken)
```

Comment:
Alright, spicey idea. Are we able to normalize this client side delegation logic to be re-used by multiple features where the the "delegated xyz" types implement maybe some base-type that says, "i'm of request X at position Y and here's some extra data for Z"?

Asking because I feel like we've wrote this boiler plate code a few times now in this file.

Reply:
I did think about this, and for background, I hate custom message target, I hate that I have to define an LSP method name, add a method with an attribute and do all of the things that make it look like an LSP call, and then its not an LSP call. Hate it!

I do think there is definitely a generic helper method here that I could write here, which uses generics and at least does the "find the virtual document, language server name, etc." for us. There is about 50 lines of code in this method, and I'd say 10 of it are interesting (the creation of RenameParams).

What I would really want, conceptually, is to go further, and do something like having a generic custom message target method that handles all of it, so I didn't have to touch this file at all. It is definitely something I would support doing, but is be more questionable as it would have to do odd JSON serialization things, which could be a perf issue.

In my ideal world I'd have something like:

```
internal class RenameEndpoint : DelegatingEndpoint
{
	public override Task<T> HandleRazor();
	public override Task<(string methodName, T? request)> GetCSharpDelegatingRequest();
	public override Task<(string methodName, T? request)> GetHtmlDelegatingRequest();
}
```

The unanswered question I have, similar to one I had above, is whether its worth it. My approach with this rename stuff was that this is the 2nd ever single server handler, and its _vastly_ different to completion. I don't want to standardize this approach, if the completion approach makes more sense, so for me, reviewing these sorts of questions after Andrew has done Hover, for example, and doing a clean up pass, makes more sense.

Reply:
> My approach with this rename stuff was that this is the 2nd ever single server handler, and its vastly different to completion.

Ya totally. I thinnkkkkkk completion may be a standout here just in general (because it needs to support provisional completion). I'll be curious if we end up needing to do more complex logic for other requests

> reviewing these sorts of questions after Andrew has done Hover, for example, and doing a clean up pass, makes more sense.

Sure that's totaly fair!

> So for example, since you mentioned it, does the delegation handle mapping positions? Or does each endpoint still do that? Do method types handle positions in the same way? What about ranges? Edits? etc. etc.

Not sure I fully understand the qs. We may want to sync-up ofline on this one